storage/storage_image.go

```diff
@@ -90,309 +95,654 @@ func newImageSource(imageRef storageReference) (*storageImageSource, error) {
 	return image, nil
 }
 
-// newImageDestination sets us up to write a new image.
+// Reference returns the image reference that we used to find this image.  Since we're officially
+// unnamed, return a reference with no naming information.
+func (s storageUnnamedImageReader) Reference() types.ImageReference {
+	return newReference(s.transport, "", s.ID, nil, "", "")
+}
+
+// Close cleans up any resources we tied up while reading the image.
+func (s storageUnnamedImageReader) Close() error {
+	return nil
+}
+
+// GetBlob reads the data blob or filesystem layer which matches the digest and size, if given.
+func (s *storageUnnamedImageReader) GetBlob(info types.BlobInfo) (rc io.ReadCloser, n int64, err error) {
+	rc, n, _, err = s.getBlobAndLayerID(info, false)
+	return rc, n, err
+}
+
+// GetBlob reads the data blob or filesystem layer which matches the digest and size, if given, and returns
+// the layer ID for the layer, if it was a layer.
+func (s *storageUnnamedImageReader) getBlobAndLayerID(info types.BlobInfo, decompress bool) (rc io.ReadCloser, n int64, layerID string, err error) {
+	var layer storage.Layer
+	var diffOptions *storage.DiffOptions
+	// We need a valid digest value.
+	err = info.Digest.Validate()
+	if err != nil {
+		return nil, -1, "", err
+	}
+	// Check if the blob corresponds to a diff that was used to initialize any layers, either
+	// before or after decompression, since we don't care.
+	uncompressedLayers, err := s.transport.store.LayersByUncompressedDigest(info.Digest)
+	compressedLayers, err := s.transport.store.LayersByCompressedDigest(info.Digest)
+	// If it's not a layer, then it must be a data item.
+	if len(uncompressedLayers) == 0 && len(compressedLayers) == 0 {
+		b, err := s.transport.store.ImageBigData(s.ID, info.Digest.String())
+		if err != nil {
+			return nil, -1, "", err
+		}
+		r := bytes.NewReader(b)
+		logrus.Debugf("exporting opaque data as blob %q", info.Digest.String())
+		return ioutil.NopCloser(r), int64(r.Len()), "", nil
+	}
+	// Step through each list of matching layers.  Tests may want to verify that if we have multiple layers
+	// which claim to have the same contents, that we actually do have multiple layers, otherwise we could
+	// just go ahead and use the first one every time.
+	i := s.layerPosition[info.Digest]
+	s.layerPosition[info.Digest] = i + 1
+	if len(uncompressedLayers) > 0 {
+		layer = uncompressedLayers[i%len(uncompressedLayers)]
+	} else {
+		layer = compressedLayers[i%len(compressedLayers)]
+	}
+	layerID = layer.ID
+	// If we've been told to force decompression, do so.
+	if decompress {
+		noCompression := archive.Uncompressed
+		diffOptions = &storage.DiffOptions{
+			Compression: &noCompression,
```

Comment:
Lately, `layerStore.Diff` only compresses a layer diff if it was originally applied in compressed form.  The files that comprise the layer are stored on disk in uncompressed form, of course, but a reader would historically ask for it using the digest from the unmodified manifest, so it was doing the best that it could by defaulting to using the same type of compression which was originally used for the layer.

A `storageImageSource` always asks for compression to be skipped, and since we're removing the ability to read layers using compressed digests from `storageUnnamedImageReader`, we don't need to bother with having it do things differently, so the `GetBlob` logic can be simplified.